### DISCUSSION TITLE
How best to make the codegen understand a custom directive?

### DISCUSSION THREAD

**Discussion by aprilrd** [dotansimha/graphql-code-generator#5676] [Q&A] [answered]
Let me start by expressing my gratitude for this set of awesome tools! I was especially blown away by "typescript-react-apollo" with its custom hooks.

Internally, we have a directive called `nonNull` on `FIELD`, which makes our gateway treat a nullable field as if it is a non-nullable field. This directive helps the client not to deal with the partial data due to engineering/design complexity.

I'd love to enable the generated types to type a field with this directive as required correctly. But I am not sure where to start for this behavior. Could I get a pointer on where to begin?

For reference, here is the definition of the directive:
```
directive @nonNull() on FIELD
```

**Comment by dotansimha** [comment, marked as answer]
I can recommend you to add a transformation layer that would change the schema according to your needs, based on the spec. So `@nonNull` can change a field to be non-nullable. This can be done with a custom schema loader: https://graphql-code-generator.com/docs/getting-started/schema-field#custom-schema-loader

**Reply by aprilrd** [reply]
I appreciate the quick response. But doesn't the schema loader operate at the schema level? The `@nonNull` directive is at the document level. I am not sure how to use the loader to affect the generated types of documents.

**Reply by dotansimha** [reply]
Oh, I see now. So I'm not sure that we have a way to support it at the moment. We had a thought to add more lifecycle hooks to the generation process, but it's not there yet.

**Reply by aprilrd** [reply]
Gotcha. When I looked at the plugin codes, I realized there wasn't a way to extend the functionality of a plugin. So lifecycle methods would be great! Or maybe a "custom document loader" works for my use case.

Regardless, thank you for answering my question.

**Comment by jamiter** [comment]
@aprilrd, ever found a way to modify the typing based on custom directives?

I'm personally looking for a way to make an `@auth` directive guarantee there is a `userId` in the context, where it's optional right now. This would prevent a lot of "if userId is string" checks in resolvers, because the directive guarantees this.

**Reply by darkbasic** [reply]
@jamiter did you find any way to implement this?

@dotansimha I'm looking for a way to change the context type from `IContext` to `WithRequired<IContext, "user" | "userId">` whenever the [accounts.js `@auth` directive](https://github.com/accounts-js/accounts/blob/master/modules/module-core/src/utils/authenticated-directive.ts) is being used in order to avoid tons of avoidable `if (user !== undefined)` in the resolvers.
```ts
export type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };
```

**Reply by jamiter** [reply]
@darkbasic, I haven't found a solution yet. Maybe @dotansimha can let us know if the required support for this feature has been added. Or send us in the right direction.

If it requires adding or modifying a current plugin, that's fine. If I find the time I could make a stab at it. I just wonder what the best approach would be.

**Reply by moetelo** [reply]
For people who come across this, see [directiveContextTypes field](https://github.com/dotansimha/graphql-code-generator/issues/7505).
